### DISCUSSION TITLE
Unable to build ISIS in ubuntu

### DISCUSSION THREAD

**Discussion by NivedithaCV** [DOI-USGS/ISIS3#5274] [Q&A]
**Issue Summary:**
I am attempting to build the ISIS software on Ubuntu, but I'm encountering an error that is preventing the successful build process. The error occurs in the `AbstractTreeModel.cpp` file of the `CnetEditorWidget` component.

**Error Message:**
In file included from /home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.cpp:9:
/home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.h:182:33: error: expected template-name before '<' token
  182 |           : public std::function<bool(AbstractTreeItem *const &)> {
      |                                 ^
/home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.h:182:33: error: expected '{' before '<' token
/home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.h:182:33: error: expected unqualified-id before '<' token
/home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.cpp: In member function 'void Isis::AbstractTreeModel::applyFilter()':
/home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.cpp:466:50: error: invalid use of incomplete type 'class Isis::AbstractTreeModel::FilterFunctor'
  466 |             FilterFunctor(m_localFilterWidgetCopy),
      |                                                  ^
In file included from /home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.cpp:9:
/home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.h:181:13: note: forward declaration of 'class Isis::AbstractTreeModel::FilterFunctor'
  181 |       class FilterFunctor
      |             ^~~~~~~~~~~~~
/home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.cpp:467:29: error: incomplete type 'Isis::AbstractTreeModel::FilterFunctor' used in nested name specifier
  467 |             &FilterFunctor::updateTopLevelLinks,
      |                             ^~~~~~~~~~~~~~~~~~~
/home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.cpp: At global scope:
/home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.cpp:591:21: error: invalid use of incomplete type 'class Isis::AbstractTreeModel::FilterFunctor'
  591 |     FilterWidget *fw) : m_filter(fw) {
      |                     ^
In file included from /home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.cpp:9:
/home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.h:181:13: note: forward declaration of 'class Isis::AbstractTreeModel::FilterFunctor'
  181 |       class FilterFunctor
      |             ^~~~~~~~~~~~~
/home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.cpp:590:3: error: 'class Isis::AbstractTreeModel::FilterFunctor' is private within this context
  590 |   AbstractTreeModel::FilterFunctor::FilterFunctor(
      |   ^~~~~~~~~~~~~~~~~
In file included from /home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.cpp:9:
/home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.h:181:13: note: declared private here
  181 |       class FilterFunctor
      |             ^~~~~~~~~~~~~
/home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.cpp:595:77: error: invalid use of incomplete type 'class Isis::AbstractTreeModel::FilterFunctor'
  595 |   AbstractTreeModel::FilterFunctor::FilterFunctor(FilterFunctor const &other) {
      |                                                                             ^
In file included from /home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.cpp:9:
/home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.h:181:13: note: forward declaration of 'class Isis::AbstractTreeModel::FilterFunctor'
  181 |       class FilterFunctor
      |             ^~~~~~~~~~~~~
/home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.cpp:595:3: error: 'class Isis::AbstractTreeModel::FilterFunctor' is private within this context
  595 |   AbstractTreeModel::FilterFunctor::FilterFunctor(FilterFunctor const &other) {
      |   ^~~~~~~~~~~~~~~~~
In file included from /home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.cpp:9:
/home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.h:181:13: note: declared private here
  181 |       class FilterFunctor
      |             ^~~~~~~~~~~~~
/home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.cpp:600:52: error: invalid use of incomplete type 'class Isis::AbstractTreeModel::FilterFunctor'
  600 |   AbstractTreeModel::FilterFunctor::~FilterFunctor() {
      |                                                    ^
In file included from /home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.cpp:9:
/home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.h:181:13: note: forward declaration of 'class Isis::AbstractTreeModel::FilterFunctor'
  181 |       class FilterFunctor
      |             ^~~~~~~~~~~~~
/home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.cpp:600:3: error: 'class Isis::AbstractTreeModel::FilterFunctor' is private within this context
  600 |   AbstractTreeModel::FilterFunctor::~FilterFunctor() {
      |   ^~~~~~~~~~~~~~~~~
In file included from /home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.cpp:9:
/home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.h:181:13: note: declared private here
  181 |       class FilterFunctor
      |             ^~~~~~~~~~~~~
/home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.cpp:605:36: error: invalid use of incomplete type 'class Isis::AbstractTreeModel::FilterFunctor'
  605 |     AbstractTreeItem *const &item) const {
      |                                    ^~~~~
In file included from /home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.cpp:9:
/home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.h:181:13: note: forward declaration of 'class Isis::AbstractTreeModel::FilterFunctor'
  181 |       class FilterFunctor
      |             ^~~~~~~~~~~~~
/home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.cpp:604:8: error: 'class Isis::AbstractTreeModel::FilterFunctor' is private within this context
  604 |   bool AbstractTreeModel::FilterFunctor::operator()(
      |        ^~~~~~~~~~~~~~~~~
In file included from /home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.cpp:9:
/home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.h:181:13: note: declared private here
  181 |       class FilterFunctor
      |             ^~~~~~~~~~~~~
/home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.cpp:612:73: error: invalid use of incomplete type 'class Isis::AbstractTreeModel::FilterFunctor'
  612 |   AbstractTreeModel::FilterFunctor::operator=(FilterFunctor const &other) {
      |                                                                         ^
In file included from /home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.cpp:9:
/home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.h:181:13: note: forward declaration of 'class Isis::AbstractTreeModel::FilterFunctor'
  181 |       class FilterFunctor
      |             ^~~~~~~~~~~~~
/home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.cpp:611:22: error: 'class Isis::AbstractTreeModel::FilterFunctor' is private within this context
  611 |   AbstractTreeModel::FilterFunctor &
      |                      ^~~~~~~~~~~~~
In file included from /home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.cpp:9:
/home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.h:181:13: note: declared private here
  181 |       class FilterFunctor
      |             ^~~~~~~~~~~~~
/home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.cpp:621:29: error: invalid use of incomplete type 'class Isis::AbstractTreeModel::FilterFunctor'
  621 |     AbstractTreeItem *item) const {
      |                             ^~~~~
In file included from /home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.cpp:9:
/home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.h:181:13: note: forward declaration of 'class Isis::AbstractTreeModel::FilterFunctor'
  181 |       class FilterFunctor
      |             ^~~~~~~~~~~~~
/home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.cpp:620:8: error: 'class Isis::AbstractTreeModel::FilterFunctor' is private within this context
  620 |   void AbstractTreeModel::FilterFunctor::filterWorker(
      |        ^~~~~~~~~~~~~~~~~
In file included from /home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.cpp:9:
/home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.h:181:13: note: declared private here
  181 |       class FilterFunctor
      |             ^~~~~~~~~~~~~
/home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.cpp:677:34: error: invalid use of incomplete type 'class Isis::AbstractTreeModel::FilterFunctor'
  677 |     AbstractTreeItem *const &item) {
      |                                  ^
In file included from /home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.cpp:9:
/home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.h:181:13: note: forward declaration of 'class Isis::AbstractTreeModel::FilterFunctor'
  181 |       class FilterFunctor
      |             ^~~~~~~~~~~~~
/home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.cpp:675:8: error: 'class Isis::AbstractTreeModel::FilterFunctor' is private within this context
  675 |   void AbstractTreeModel::FilterFunctor::updateTopLevelLinks(
      |        ^~~~~~~~~~~~~~~~~
In file included from /home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.cpp:9:
/home/niser/Desktop/ISIS3/isis/src/qisis/objs/CnetEditorWidget/AbstractTreeModel.h:181:13: note: declared private here
  181 |       class FilterFunctor
      |             ^~~~~~~~~~~~~
make[2]: *** [objects/CMakeFiles/isis.dir/build.make:10382: objects/CMakeFiles/isis.dir/qisis/objs/CnetEditorWidget/AbstractTreeModel.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:5052: objects/CMakeFiles/isis.dir/all] Error 2
make: *** [Makefile:146: all] Error 2

**Steps to Reproduce:**
Build was done as instructed in [https://github.com/DOI-USGS/ISIS3/wiki/Developing-ISIS3-with-cmake#building-isis3](url) using Make

**Expected Behavior:**
I expect the ISIS software to build successfully without encountering the mentioned error.

**Additional Context:**
I have reviewed the main CMake file and the relevant CMake files in the `cmake` directories, but I haven't been able to identify the source of the error. 

**Environment:**
- Operating System: Ubuntu 20.04.6 LTS
- ISIS Version: 7.2.0

I kindly request assistance in diagnosing and resolving this issue. If you have any insights, suggestions, or advice on how to address the error, I would greatly appreciate it.

Thank you

**Comment by acpaquette** [comment]
Hi @NivedithaCV I would double check that your system has all of its packages up to date, running `sudo apt update`. If you are still unable to build after that it may be necessary to run `sudo apt install build-essential` this will install various build tools like gcc and make. Hopefully that pushes things forward!

**Reply by NivedithaCV** [reply]
I am getting the same error. It didn't fix the problem

**Reply by acpaquette** [reply]
I can spin up a ubuntu machine on my end and try to reproduce this behavior there. Also is there a particular reason you want to build against 7.2.0? Something else to try is building against the dev branch off of the repo, rather than the 7.2.0 branch specifically.

**Reply by NivedithaCV** [reply]
I have used an older clone, and it has been successfully built.

**Reply by acpaquette** [reply]
@NivedithaCV This has been identified as an issue with ISIS and should be fixed by #5278

**Comment by acpaquette** [comment]
Closing this as #5278 was merged
